pkg/graveler/graveler.go

```diff
@@ -1841,6 +1873,46 @@ func (g *Graveler) safeBranchWrite(ctx context.Context, log logging.Logger, repo
 	return nil
 }
 
+func (g *Graveler) UpdateObjectMetadata(ctx context.Context, repository *RepositoryRecord, branchID BranchID, key Key, update ValueUpdateFunc, opts ...SetOptionsFunc) error {
+	isProtected, err := g.protectedBranchesManager.IsBlocked(ctx, repository, branchID, BranchProtectionBlockedAction_STAGING_WRITE)
+	if err != nil {
+		return err
+	}
+	if isProtected {
+		return ErrWriteToProtectedBranch
+	}
+
+	options := NewSetOptions(opts)
+	if repository.ReadOnly && !options.Force {
+		return ErrReadOnlyRepository
+	}
+
+	log := g.log(ctx).WithFields(logging.Fields{"key": key, "operation": "update_user_metadata"})
+
+	err = g.safeBranchWrite(ctx, log, repository, branchID, safeBranchWriteOptions{MaxTries: options.MaxTries}, func(branch *Branch) error {
+		value, onStaging, err := g.getFromBranchWithStaging(ctx, repository, branch, key)
+		if err != nil && !errors.Is(err, ErrNotFound) {
+			return fmt.Errorf("get existing entry to update user metadata: %w", err)
+		}
+
+		return g.StagingManager.Update(ctx, branch.StagingToken, key, func(currentValue *Value) (*Value, error) {
+			if currentValue == nil {
+				if onStaging {
```

Comment:
Do we really need this piece of code?
Considering this will just fail during the update due to the change on the branch?
This can also result in inconsistent behavior:
If we catch the deletion before this part we will return `ErrNotFound` but if the delete occurs after this check we will get `PreconditionFalied`.
This might also cause confusion when debugging

Reply:
Well due to the code changes it is not relevant but, consider the two timings:
1. Object is deleted from staging before this line - we will return ErrNotFound
2. Object is initially read (with pred) at the beginning of `g.StagingManager.Update` then immediately deleted and we use the setMsgIf using the pred - and we get `ErrPredicateFailed`
In both scenarios the object was deleted in staging during UpdateMetadata

Reply:
Thanks!

I think that you were right and still are, and that this still happens.  But only during a race.  It is legit to fail an update if it occurs concurrently with a deletion.  Returning ErrNotFound is the best way, which serializes DELETE before UPDATE.  But returning ErrPredicateFailed is equally valid - it indicates that Graveler was unable to serialize the operations.

I doubt this will matter in practice.  If it does, we will need to add a retry.